src/com/palmergames/bukkit/towny/tasks/DrawSmokeTask.java

```diff
@@ -22,13 +40,71 @@ public DrawSmokeTask(Towny plugin) {
 	public void run() {
 		for (Player player : Bukkit.getOnlinePlayers()) {
 			Resident resident = TownyAPI.getInstance().getResident(player);
-			if (resident == null || !resident.hasMode("constantplotborder"))
+			if (resident == null)
+				continue;
+
+			if (resident.hasMode("constantplotborder")) {
+				WorldCoord wc = new WorldCoord(player.getWorld().getName(), Coord.parseCoord(player.getLocation()));
+				CellBorder cellBorder = BorderUtil.getPlotBorder(wc);
+				
+				cellBorder.runBorderedOnSurface(1, 2, DrawSmokeTaskFactory.showToPlayer(player, DrawSmokeTaskFactory.getAffiliationColor(resident, wc)));
+				continue;
+			}
+
+			if (resident.hasMode("townborder")) {
+				Town town = TownyAPI.getInstance().getTown(player.getLocation());
+				if (town == null)
+					continue;
+
+				List<CellBorder> cellBorders = getCellBorders(new TownWorldPair(town, player.getWorld()));
+				if (cellBorders == null)
+					continue;
+
+				Color color = DrawSmokeTaskFactory.getAffiliationColor(resident, cellBorders.get(0));
+				cellBorders.forEach(cb -> cb.runBorderedOnSurface(1, 2, DrawSmokeTaskFactory.showToPlayer(player, color)));
 				continue;
+			}
+		}
+	}
+
+	@Nullable
+	private List<CellBorder> getCellBorders(TownWorldPair pair) {
+		List<CellBorder> cellBorders = null;
+		try {
+			// Keyed to a String so that it is always a single memory address.
+			cellBorders = cellBorderCache.get(pair.toString(), new Callable<List<CellBorder>>() {
+				@Override
+				public List<CellBorder> call() {
+					return pair.getCellBordersForTownInWorld();
+				}
+			});

```

Comment:
Callable is a functional interface so this can be reduced down to just `get(String, pair::getCellBordersForTownInWorld);`